src/advanced_inv.cpp

```diff
@@ -1450,27 +1505,21 @@ void advanced_inventory::display()
                     }
                 }
             } else {
+                bool need_to_redraw = false;
                 // from map/vehicle: add the item to the destination, if that worked,
                 // remove it from the source, else continue.
-                // TODO: move partial stack with items that are not counted by charges
-                item new_item( *sitem->it );
-                if( by_charges ) {
+                // create a new copy of the old item being manipulated
+                item new_item(*sitem->items.front());
+                if(by_charges) {
+                    // set the new item's charge amount
                     new_item.charges = amount_to_move;
+                    // `amount_to_move' will be `true' if the item needs to be removed
+                    amount_to_move = sitem->items.front()->reduce_charges(amount_to_move);
                 }
-                if( !add_item( destarea, new_item ) ) {
-                    // make sure to redraw the info window, so 
-                    // the user can see why it wasn't equipped
-                    if(destarea == AIM_WORN) {
-                        redraw = 1;
-                    }
-                    continue;
-                }
-
-                if( by_charges && amount_to_move < sitem->it->charges ) {
-                    sitem->it->charges -= amount_to_move;
-                } else {
-                    remove_item( *sitem );
-                }
+                need_to_redraw = !add_item(destarea, new_item, (by_charges) ? 1 : amount_to_move);
+                remove_item(*sitem, amount_to_move);
+                // redraw the screen if moving to AIM_WORN, so we can see that it didn't work
+                redraw = destarea == AIM_WORN && need_to_redraw;
```

Comment:
The logic in this block has been changed: Previously it would call `add_item`, and bail out if that failed. In that case nothing had been changed (neither in the source nor in the target area). If adding succeeded, `remove_item` would get called, which can not fail.

Now you're reducing the charges *before* calling `add_item`. What if `add_item` fails? `remove_item` is called anyway. The result of `add_item` is ignored (at least for the behavior, it's only used to trigger redrawing the screen).

So if I try to move a stack of (mismatching) ammo to a vehicle turret and `add_item` fails (as expected), the "moved" charges are removed anyway. Not good.

----

Another (this time testable) bug: get 3 armor items of the same type. Open the AIM, select inventory and worn pane. Now move all items from the inventory to the worn pane. The first two may work, but the third item can not be worn (only 2 items of each type can be worn), but the items is removed from the inventory anyway (it ends up on the ground).

Now try the same but have the items on the ground (and try to wear them from there). The items will no vanish.